### PR TITLE
Feature/feature/assignments performance optimizations r3

### PR DESCRIPTION
**Pull request by bargenilesh**
Stability fixes after endurance test:
	-	Fix for ConcurrentModificationExcpetion while accessing BucketList
  	-	More efficient way of refreshing Assignment metadata cache to avoid frequent spike.
  	-	Added a limit of 2gb on the metadata cache
  	-	Make ehcache.xml part of conf directory	
        -	Added exception handling at AssignmentResource level to log any uncaught exception.
  	-	Some relevant assignment metadata cache logging changes.

### CODE REVIEW

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java [hunk 1]
```diff
@@ -50,8 +52,13 @@
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 
+import static com.intuit.wasabi.assignment.cache.AssignmentsMetadataCache.CACHE_NAME.APP_NAME_N_PAGE_TO_EXPERIMENTS_CACHE;
+import static com.intuit.wasabi.assignment.cache.AssignmentsMetadataCache.CACHE_NAME.APP_NAME_TO_EXPERIMENTS_CACHE;
```
Comment (mans4singh) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java]: It might be simpler to just import CACHE_NAME.*;

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java [hunk 2]
```diff
@@ -289,7 +259,14 @@ public BucketList getBucketList(Experiment.ID expId) {
             bucketList = (BucketList) val.getObjectValue();
         }
 
-        return isNull(bucketList)?new BucketList():bucketList;
+        return isNull(bucketList)?new BucketList():makeCopy(bucketList);
+    }
+
+    private BucketList makeCopy(BucketList bucketList) {
+        List<Bucket> iBucketList = new ArrayList<>(bucketList.getBuckets());
```
Comment (mans4singh) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java]: What's the reason for shallow copy here ?

File: bin/fpm.sh [hunk 3]
```diff
@@ -93,6 +93,7 @@ for module in "$modules"; do
     modules/api/target/classes/api.properties=${home}/conf/api.properties\
     modules/api/target/generated/swagger-ui/=${home}/content/ui/dist/swagger/swaggerjson\
     modules/assignment/target/classes/assignment.properties=${home}/conf/assignment.properties\
+    modules/assignment/target/classes/ehcache.xml=${home}/conf/ehcache.xml\
```
Comment (iizrailevsky) [inline bin/fpm.sh]: Do we need a corresponding update on wasabi-intuit [cc: @jcwuzoegiver] to include ehcahce.xml there?

File: modules/assignment/src/main/resources/ehcache.xml [hunk 4]
```diff
@@ -19,7 +19,9 @@
          xsi:noNamespaceSchemaLocation="http://ehcache.org/ehcache.xsd"
          updateCheck="true"
          monitoring="autodetect"
-         dynamicConfig="true">
+         dynamicConfig="true"
+         maxBytesLocalHeap="2G"
```
Comment (iizrailevsky) [inline modules/assignment/src/main/resources/ehcache.xml]: Thanks for setting a cache memory limit to ensure we don't run out of memory! 👍

**Comment by iizrailevsky** [comment]
Looks good 👍  :shipit:

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java (hunk 2 above)
Comment (jcwuzoegiver) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java]: same question. Why making a copy of bucket list?

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheRefreshTask.java [hunk 5]
```diff
@@ -49,7 +49,8 @@ public AssignmentsMetadataCacheRefreshTask(AssignmentsMetadataCache metadataCach
     @Override
     public void run() {
         try {
-            LOGGER.info("AssignmentsMetadataCache refresh started at = {}", timeService.getCurrentTime());
+            Date startTime = timeService.getCurrentTime();
+            LOGGER.debug("AssignmentsMetadataCache refresh started at = {}", startTime);
```
Comment (jcwuzoegiver) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheRefreshTask.java]: why LOGGER.debug() and then using info() later at line 65 when logging the total time it takes to refresh?

File: modules/assignment/src/main/resources/ehcache.xml (hunk 4 above)
Comment (jcwuzoegiver) [inline modules/assignment/src/main/resources/ehcache.xml]: is 2G based on the type of ec2 instance we use to deploy wasabi app?

File: bin/fpm.sh (hunk 3 above)
Comment (jcwuzoegiver) [inline bin/fpm.sh]: @iizrailevsky i believe so. need to do the same for wasabi-intuit's fpm.sh

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java (hunk 2 above)
Comment (jcwuzoegiver) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java]: @bargenilesh why is there ConcurrentModificationExcpetion for  BucketList? isn't the time interval 5 mins? how come multiple threads are trying to modify bucket list?

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java (hunk 1 above)
Comment (bargenilesh) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java]: Yes, you are right, it is just what IntelliJ formatter...

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java (hunk 2 above)
Comment (bargenilesh) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheImpl.java]: Making a copy of BucketList to avoid ConcurrentModificationException... At some point during the assignment logic, the List<Bucket> object inside this BucketList is getting sorted, I believe that's the reason of ConcurrentModificationException. 

AssignmentImpl.selectBucket(List<Bucket> buckets) Line#1189.

Provided comments to this method...

File: bin/fpm.sh (hunk 3 above)
Comment (bargenilesh) [inline bin/fpm.sh]: Already done and created a separate PR for wasabi-intuit

File: modules/assignment/src/main/resources/ehcache.xml (hunk 4 above)
Comment (bargenilesh) [inline modules/assignment/src/main/resources/ehcache.xml]: 2gb is just the best guess higher limit.. We currently run app instances with 8gb of memory, out of which consistently 1.1gb is being used...

File: modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheRefreshTask.java (hunk 5 above)
Comment (bargenilesh) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheRefreshTask.java]: In PROD, App runs at the INFO level. And I wanted to just print a single line highlighting that cache has been refreshed, so kept a single log with INFO level with the time taken to refresh.

This is based on the endurance test experiment where 3 lines were being logged at the cache refresh event, that makes it difficult to understand/track/validate behavior.

I case of any further debugging, we can explicitly enable DEBUG log for this class, which would be rare case scenario mostly in PRE-PROD or LOCAL environment.

Comment (jcwuzoegiver) [inline modules/assignment/src/main/java/com/intuit/wasabi/assignment/cache/impl/AssignmentsMetadataCacheRefreshTask.java]: ok

**Comment by jcwuzoegiver** [comment]
Looks good after incorporating the feedback. :shipit:
